src/PhenX.EntityFrameworkCore.BulkInsert.Oracle/OracleDialectBuilder.cs

```diff
@@ -0,0 +1,136 @@
+        OnConflictOptions? onConflict = null)
+    {
+        var q = new StringBuilder();
+
+        // Merge handling
+        if (onConflict is OnConflictOptions<T> onConflictTyped)
+        {
+            IEnumerable<string> matchColumns;
+            if (onConflictTyped.Match != null)
+            {
+                matchColumns = GetColumns(target, onConflictTyped.Match);
+            }
+            else if (target.PrimaryKey.Count > 0)
+            {
+                matchColumns = target.PrimaryKey.Select(x => x.QuotedColumName);
+            }
+            else
+            {
+                throw new InvalidOperationException("Table has no primary key that can be used for conflict detection.");
+            }
+
+            q.AppendLine($"MERGE INTO {target.QuotedTableName} AS TARGET");
+
+            q.Append("USING (SELECT ");
+            q.AppendColumns(insertedColumns);
+            q.Append($" FROM {source}) AS SOURCE (");
+            q.AppendColumns(insertedColumns);
+            q.AppendLine(")");
+
+            q.Append("ON ");
+            q.AppendJoin(" AND ", matchColumns, (b, col) => b.Append($"TARGET.{col} = SOURCE.{col}"));
+            q.AppendLine();
+
+            if (onConflictTyped.Update != null)
+            {
+                var columns = target.GetColumns(false);
+
+                q.AppendLine("WHEN MATCHED THEN UPDATE SET ");
+                q.AppendJoin(", ", GetUpdates(context, target, columns, onConflictTyped.Update));
+                q.AppendLine();
+            }
+
+            q.Append("WHEN NOT MATCHED THEN INSERT (");
+            q.AppendColumns(insertedColumns);
+            q.AppendLine(")");
+
+            q.Append("VALUES (");
+            q.AppendJoin(", ", insertedColumns, (b, col) => b.Append($"SOURCE.{col.QuotedColumName}"));
+            q.AppendLine(")");
+
+            if (returnedColumns.Count != 0)
+            {
+                q.Append("OUTPUT ");
+                q.AppendJoin(", ", returnedColumns, (b, col) => b.Append($"INSERTED.{col.QuotedColumName} AS {col.QuotedColumName}"));
+                q.AppendLine();
+            }
+        }
+
+        // No conflict handling
+        else
+        {
+            q.Append($"INSERT INTO {target.QuotedTableName} (");
+            q.AppendColumns(insertedColumns);
+            q.AppendLine(")");
+            q.Append("SELECT ");
+            q.AppendColumns(insertedColumns);
+            q.AppendLine();
+            q.Append($"FROM {source}");
+            q.AppendLine();
+
+            if (returnedColumns.Count != 0)
+            {
+                q.Append("RETURNING ");
+                q.AppendJoin(", ", returnedColumns, (b, col) => b.Append(col.QuotedColumName));
+                q.Append(" INTO ");
+                q.AppendJoin(", ", returnedColumns, (b, col) => b.Append($":{col.ColumnName}"));
+                q.AppendLine();
+            }
+        }
+
+        q.AppendLine(";");
+
+        var result = q.ToString();
```

Comment:
Have you not got rid of this syntax in another PR

aka

return q.ToString();